crates/chia-bls/src/cached_bls.rs

```diff
@@ -0,0 +1,294 @@
+                let aug_hash: Signature = hash_to_g2(&aug_msg);
+
+                let pk_parsed: &mut PublicKey = pk_bytes_to_g1
+                    .entry(pks.borrow()[i])
+                    .or_insert_with(|| PublicKey::from_bytes(&pks.borrow()[i]).unwrap());
+
+                let pairing: GTElement = aug_hash.pair(pk_parsed);
+                let mut hasher = Sha256::new();
+                hasher.update(&aug_msg);
+                let h: Bytes32 = hasher.finalize().into();
+                self.cache.put(h, pairing.clone());
+                ret.push(pairing);
+            }
+        }
+
+        ret
+    }
+
+    pub fn aggregate_verify(
+        &mut self,
+        pks: &Vec<Bytes48>,
+        msgs: &Vec<Vec<u8>>,
+        sig: &Signature,
+        force_cache: bool,
+    ) -> bool {
+        let mut pairings: Vec<GTElement> = self.get_pairings(pks, msgs, force_cache);
+        if pairings.is_empty() {
+            let mut data = Vec::<(PublicKey, Vec<u8>)>::new();
+            for (pk, msg) in pks.iter().zip(msgs.iter()) {
+                let pk = PublicKey::from_bytes_unchecked(pk).unwrap();
+                data.push((pk.clone(), msg.clone()));
+            }
+            let res: bool = agg_ver(sig, data);
+            return res;
+        }
+        let pairings_prod = pairings.pop(); // start with the first pairing
+        match pairings_prod {
+            Some(mut prod) => {
+                for p in pairings.iter() {
+                    // loop through rest of list
+                    prod *= &p;
+                }
+                prod == sig.pair(&PublicKey::generator())
+            }
+            _ => pairings.is_empty(),
+        }
+    }
+}
+
+// Python Functions
+#[cfg(feature = "py-bindings")]
+#[pymethods]
+impl BLSCache {
+    #[new]
+    pub fn init() -> Self {
+        Self::generator(None)
+    }
+
+    #[staticmethod]
+    #[pyo3(name = "generator")]
+    pub fn py_generator(size: Option<&PyInt>) -> Self {
+        match size {
+            Some(s) => {
+                let usize_value: usize = s.extract::<usize>().unwrap();
+                Self::generator(Some(usize_value))
+            }
+            None => Self::generator(None),
+        }
+    }
+
+    #[pyo3(name = "aggregate_verify")]
+    pub fn py_aggregate_verify(
+        &mut self,
+        pks: &PyList,
+        msgs: &PyList,
+        sig: &Signature,
+        force_cache: &PyBool,
+    ) -> PyResult<bool> {
+        let pks_r: Vec<Bytes48> = pks
+            .iter()
+            .map(|item| item.extract::<Bytes48>())
+            .collect::<PyResult<_>>()?;
+        let msgs_r: Vec<Vec<u8>> = msgs
+            .iter()
+            .map(|item| item.extract::<Vec<u8>>())
+            .collect::<PyResult<_>>()?;
+        let force_cache_bool = force_cache.extract::<bool>()?;
+        Ok(self.aggregate_verify(&pks_r, &msgs_r, sig, force_cache_bool))
+    }
+
+    #[pyo3(name = "len")]
```

Comment:
I think you can implement magic methods like `__len__`, but I suppose it doesn't matter much.